lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp

```diff
@@ -0,0 +1,115 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Conversion/TTNNToEmitC/TTNNToEmitC.h"
+
+#include "../PassDetail.h"
+
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Dialect/EmitC/IR/EmitC.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/MLIRContext.h"
+
+#include "ttmlir/Dialect/TTNN/IR/TTNN.h"
+#include "ttmlir/Dialect/TTNN/IR/TTNNOps.h"
+
+using namespace mlir;
+using namespace mlir::tt;
+
+namespace {
+
+class TTNNToEmitCTypeConverter : public TypeConverter {
+public:
+  TTNNToEmitCTypeConverter(MLIRContext *ctx) {
+    addConversion([](Type type) { return type; });
+    addConversion([ctx](mlir::tt::DeviceType type) -> Type {
+      return emitc::OpaqueType::get(ctx, "ttnn::Device");
+    });
+    addConversion([ctx](TensorType type) -> Type {
+      return emitc::OpaqueType::get(ctx, "ttnn::Tensor");
+    });
+  }
+};
+
+struct ConvertTTNNToEmitCPass
+    : public ttnn::ConvertTTNNToEmitCBase<ConvertTTNNToEmitCPass> {
+  void runOnOperation() override {
+    mlir::ConversionTarget target(getContext());
+
+    target.addLegalDialect<emitc::EmitCDialect>();
+    target.addIllegalDialect<ttnn::TTNNDialect>();
+    target.addLegalOp<mlir::ModuleOp>();
+
+    // Add header imports to front of module
+    //
+    {
+      auto module = getOperation();
+      OpBuilder builder(module);
+
+      builder.create<emitc::IncludeOp>(module.getLoc(), "ttnn/device.h",
+                                       /*isStandard=*/false);
+      builder.create<emitc::IncludeOp>(
+          module.getLoc(), "ttnn/operations/eltwise/binary/binary.hpp",
+          /*isStandard=*/false);
+      builder.create<emitc::IncludeOp>(
+          module.getLoc(), "ttnn/operations/core.hpp", /*isStandard=*/false);
+      builder.create<emitc::IncludeOp>(module.getLoc(),
+                                       "ttnn/operations/creation.hpp",
+                                       /*isStandard=*/false);
+      builder.create<emitc::IncludeOp>(
+          module.getLoc(),
+          "ttnn/operations/reduction/generic/generic_reductions.hpp",
+          /*isStandard=*/false);
+      builder.create<emitc::IncludeOp>(module.getLoc(),
+                                       "ttnn/operations/normalization.hpp",
+                                       /*isStandard=*/false);
+    }
+
+    // TTNN -> EmitC
+    //
+    {
+      TTNNToEmitCTypeConverter typeConverter(&getContext());
+      RewritePatternSet patterns(&getContext());
+
+      // Func dialect handling
```

Comment:
What happens if you treat the FucnOp as legal and pass it down without any conversion?

Reply:
That should work, per emitc docs, they can parse it: https://mlir.llvm.org/docs/Dialects/EmitC/

However, I think I tried earlier, and was hitting some "materialization" errors which I didn't quite understand. Let me try again.

Reply:
With the following diff
```diff
diff --git a/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp b/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp
index e42e50f..b00ad00 100644
--- a/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp
+++ b/lib/Conversion/TTNNToEmitC/TTNNToEmitCPass.cpp
@@ -15,6 +15,7 @@

 #include "ttmlir/Dialect/TTNN/IR/TTNN.h"
 #include "ttmlir/Dialect/TTNN/IR/TTNNOps.h"
+#include <mlir/Dialect/Func/IR/FuncOps.h>

 using namespace mlir;
 using namespace mlir::tt;
@@ -42,6 +43,8 @@ struct ConvertTTNNToEmitCPass
     target.addLegalDialect<emitc::EmitCDialect>();
     target.addIllegalDialect<ttnn::TTNNDialect>();
     target.addLegalOp<mlir::ModuleOp>();
+    target.addLegalOp<func::FuncOp>();
+    target.addLegalOp<func::ReturnOp>();

     // Add header imports to front of module
     //
@@ -74,20 +77,20 @@ struct ConvertTTNNToEmitCPass
       TTNNToEmitCTypeConverter typeConverter(&getContext());
       RewritePatternSet patterns(&getContext());

-      // Func dialect handling
-      //
-      populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
-          patterns, typeConverter);
-      target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
-        return typeConverter.isSignatureLegal(op.getFunctionType()) &&
-               typeConverter.isLegal(&op.getBody());
-      });
-      populateReturnOpTypeConversionPattern(patterns, typeConverter);
-      target.addDynamicallyLegalOp<func::ReturnOp>(
-          [&](func::ReturnOp op) { return typeConverter.isLegal(op); });
-      populateCallOpTypeConversionPattern(patterns, typeConverter);
-      target.addDynamicallyLegalOp<func::CallOp>(
-          [&](func::CallOp op) { return typeConverter.isLegal(op); });
+      // // Func dialect handling
+      // //
+      // populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
+      //     patterns, typeConverter);
+      // target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      //   return typeConverter.isSignatureLegal(op.getFunctionType()) &&
+      //          typeConverter.isLegal(&op.getBody());
+      // });
+      // populateReturnOpTypeConversionPattern(patterns, typeConverter);
+      // target.addDynamicallyLegalOp<func::ReturnOp>(
+      //     [&](func::ReturnOp op) { return typeConverter.isLegal(op); });
+      // populateCallOpTypeConversionPattern(patterns, typeConverter);
+      // target.addDynamicallyLegalOp<func::CallOp>(
+      //     [&](func::CallOp op) { return typeConverter.isLegal(op); });

       // TTNN -> EmitC patterns
       //
```

if I run
```bash
./build/bin/ttmlir-opt --ttir-layout --ttnn-open-device --convert-ttir-to-ttnn --convert-ttnn-to-emitc test/ttmlir/Dialect/TTNN/simple_multiply.mlir | ./build/bin/ttmlir-translate -mlir-to-cpp -allow-unregistered-dialect
```

I get

```cpp
test/ttmlir/Dialect/TTNN/simple_multiply.mlir:10:10: error: failed to legalize unresolved materialization from 'tensor<64x128xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #tt.memory_space<system>>>>' to '!emitc.opaque<"ttnn::Tensor">' that remained live after conversion
    %1 = "ttir.multiply"(%arg0, %arg1, %0) <{operandSegmentSizes = array<i32: 2, 1>, operand_constraints = [#any_device, #any_device, #any_device]}> : (tensor<64x128xf32>, tensor<64x128xf32>, tensor<64x128xf32>) -> tensor<64x128xf32>
         ^
test/ttmlir/Dialect/TTNN/simple_multiply.mlir:10:10: note: see current operation: %0 = "builtin.unrealized_conversion_cast"(%arg1) : (tensor<64x128xf32, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xf32, #tt.memory_space<system>>>>) -> !emitc.opaque<"ttnn::Tensor">
test/ttmlir/Dialect/TTNN/simple_multiply.mlir:10:10: note: see existing live user here: %10 = emitc.call_opaque "ttnn::to_memory_config"(%0, %8) : (!emitc.opaque<"ttnn::Tensor">, !emitc.opaque<"ttnn::Tensor">) -> !emitc.opaque<"ttnn::Tensor">
```

Reply:
I think I understand what's going on... If I don't have actual conversion patterns added to `RewritePatternSet`, operands of func ops won't be converted. Having these calls

```cpp
      populateFunctionOpInterfaceTypeConversionPattern<func::FuncOp>(
          patterns, typeConverter);
      populateReturnOpTypeConversionPattern(patterns, typeConverter);
      populateCallOpTypeConversionPattern(patterns, typeConverter);
```

takes care of that.

Reply:
Reminder that we can delete "PassDetail.h" header and move `GEN_PASS_DEF_CONVERTTTNNTOEMITC` step into the .cpp file.